### ISSUE TITLE
Finite-sized `r.range` calls should be considered arrays rather than streams

### ISSUE THREAD

**Issue by mlucy** [rethinkdb/rethinkdb#5063]
There's no reason anyone should ever have to write `r.range(100).coerce_to('array')`.  I think this was just an oversight rather than an intentional part of the interface.

**Comment by VeXocide** [comment]
I've often used finite-sized ranges larger than the array size limit, how should that be handled? In my opinion `r.range(100)` really shouldn't be different from `r.range(200000)`, or have `r.range(10000)` behaviour depend on the configured array size limit.

**Comment by mlucy** [comment]
Fair point.  Actually, this will go away when we fix https://github.com/rethinkdb/rethinkdb/issues/2721 anyway, so on second thought let's just not do anything about this until then.

**Comment by mlucy** [comment]
Re-opening this: #2721 is looking like it might not happen for a while, and this is still *incredibly* annoying.

**Comment by mlucy** [comment]
One thing to do would be to make `r.range` with a finite argument an array by default, but add an optarg to produce a stream.  That way the person who wants to write `r.range(200000)` (which is fairly rare) would be the one who has to do the extra typing, rather than the person who wants to write `r.range(100)` or `r.range(arr.count())`, both of which I do pretty frequently.

**Comment by mlucy** [comment]
(We could also be clever and automatically use the stream optarg if the argument is larger than the current array size limit.  I'm not sure if that's better or worse.)

**Comment by mlucy** [comment]
(Incidentally, the impetus for this was helping a user with http://stackoverflow.com/questions/34828117/rethinkdb-how-to-return-a-sliding-window-on-grouped-data .  I use `r.range` to answer a StackOverflow question at least once a month, and I literally always want it to return an array.)

**Comment by danielmewes** [comment]
I don't think we should try doing anything "smart". Arrays and streams need to be handled sufficiently differently in ReQL, that automatically switching from one to the other will only be confusing.

I really like the fact that `r.range` is a stream, but I think I'd be ok with changing it to an array by default and having an optarg for making it a stream again.
I'm not sure if I would prefer an unbounded `r.range` without the `stream` optarg to fail, or to switch to the stream return type automatically.

**Comment by VeXocide** [comment]
I've often used `r.range` to fill a table with some initial data, and there it makes more sense for it to behave as a stream. Automatically switching sounds too smart to me as well, in my opinion we should simply leave it as a stream.

**Comment by deontologician** [comment]
+1 for leaving infinite `r.range` as a stream with no optargs. I think the fact that you call it differently should be sufficient to make the distinction that we don't need an optarg.
